src/pybullet_blocks/envs/cleanup_table_env.py

```diff
@@ -0,0 +1,768 @@
+        """Get the dimensionality of nodes."""
+        return max(
+            RobotState.get_dimension(),
+            LetteredBlockState.get_dimension(),
+            BlockState.get_dimension(),
+        )
+
+    def to_observation(self) -> spaces.GraphInstance:
+        """Create graph representation of the state."""
+        # Add bin as a special node type (type 2)
+        bin_node = np.zeros(self.get_node_dimension(), dtype=np.float32)
+        bin_node[0] = 2  # Distinguish bin from robot(0) and toys(1)
+        bin_vec = self.bin_state.to_vec()
+        bin_node[1 : 1 + len(bin_vec)] = bin_vec
+
+        inner_vecs: list[NDArray] = [
+            self.robot_state.to_vec(),
+            bin_node,
+        ]
+
+        for toy_state in self.toy_states:
+            inner_vecs.append(toy_state.to_vec())
+
+        padded_vecs: list[NDArray] = []
+        for vec in inner_vecs:
+            padded_vec = np.zeros(self.get_node_dimension(), dtype=np.float32)
+            padded_vec[: len(vec)] = vec
+            padded_vecs.append(padded_vec)
+
+        arr = np.array(padded_vecs, dtype=np.float32)
+        return spaces.GraphInstance(nodes=arr, edges=None, edge_links=None)
+
+    @classmethod
+    def from_observation(
+        cls, obs: spaces.GraphInstance
+    ) -> CleanupTablePyBulletBlocksState:
+        """Build a state from a graph."""
+        robot_state: RobotState | None = None
+        bin_state: BlockState | None = None
+        toy_states: list[LetteredBlockState] = []
+
+        for node in obs.nodes:
+            if np.isclose(node[0], 0):  # Robot
+                assert robot_state is None
+                vec = node[: RobotState.get_dimension()]
+                robot_state = RobotState.from_vec(vec)
+            elif np.isclose(node[0], 2):  # Bin
+                bin_vec = node[1 : BlockState.get_dimension()]
+                bin_state = BlockState.from_vec(bin_vec)
+            elif np.isclose(node[0], 1):  # Toy (LetteredBlockState)
+                vec = node[: LetteredBlockState.get_dimension()]
+                toy_state = LetteredBlockState.from_vec(vec)
+                toy_states.append(toy_state)
+
+        assert robot_state is not None
+        assert bin_state is not None
+
+        toy_states.sort(key=lambda x: x.letter)
+
+        return cls(toy_states, bin_state, robot_state)
+
+
+class ObjaverseLoader:
+    """Helper class for loading Objaverse objects."""
+
+    def __init__(self, config: ObjaverseConfig) -> None:
+        self.config = config
+        os.makedirs(config.cache_dir, exist_ok=True)
+        self._downloaded_objects: dict[str, str] = {}
+        self._setup_ssl_context()
+        self._download_all_objects()
+
+    def get_object_half_extents(
+        self, physics_client_id: int, body_id: int
+    ) -> tuple[float, float, float]:
+        """Get actual half-extents of a loaded object."""
+        aabb_min, aabb_max = p.getAABB(body_id, physicsClientId=physics_client_id)
```

Comment:
beware that this will change depending on the pose of the object! (if it's rotated, the AABB will be different)